### ISSUE TITLE
Nuget restore and Visual Studio build produce different project.assets.json files with C++/CLI transitive references

### ISSUE THREAD

**Issue by AlexBaratheon** [NuGet/Home#13862]
### NuGet Product Used

NuGet.exe

### Product Version

- NuGet Version 6.9.1.3
- Visual Studio 2022 Version 17.8.5

### Repro Steps & Context
`nuget restore SOLUTION` and building the solution in Visual Studio produce differend `projet.asstes.json` files. Transitive referenced packages are missing in the asset file when doing nuget restore. To reproduce it, it is important to have Visual Studio closed when doing nuget restore.
 
Project dependency structure to reproduce: C# .net8 -> C++/CLI .net8 -> C# net8 -> nuget-package

In my example project: ApplicationUtils -> ToolsCoreDynLib -> Utils -> System.Collections.Immutable
 [ProjectAssetJsonTests.zip](https://github.com/user-attachments/files/17407136/ProjectAssetJsonTests.zip)

Compare the two `project.assets.json`:
![2024-10-17 08_54_01-Window](https://github.com/user-attachments/assets/bf20fbe3-d585-4ead-b41c-941afbc6fa70)
**-> ToolsCoreDynLib is missing the dependency to Utils**

### Verbose Logs

`project.assets.json` of ApplicationUtils

First part after nuget restore solution:
> {
> 
>   "version": 3,
>   "targets": {
>     "net8.0-windows7.0": {
>       "ToolsCoreDynLib/1.0.0": {
>         "type": "project",
>         "compile": {
>           "bin/placeholder/ToolsCoreDynLib.dll": {}
>         },
>         "runtime": {
>           "bin/placeholder/ToolsCoreDynLib.dll": {}
>         }
>       }
>     }
>   },
>   "libraries": {
>     "ToolsCoreDynLib/1.0.0": {
>       "type": "project",
>       "path": "../ToolsCoreDynLib/ToolsCoreDynLib.vcxproj",
>       "msbuildProject": "../ToolsCoreDynLib/ToolsCoreDynLib.vcxproj"
>     }
>   },`

  
First part after Visual Studio solution build:
>  {
> 
>   "version": 3,
>   "targets": {
>     "net8.0-windows7.0": {
>       "System.Collections.Immutable/8.0.0": {
>         "type": "package",
>         "compile": {
>           "lib/net8.0/System.Collections.Immutable.dll": {
>             "related": ".xml"
>           }
>         },
>         "runtime": {
>           "lib/net8.0/System.Collections.Immutable.dll": {
>             "related": ".xml"
>           }
>         },
>         "build": {
>           "buildTransitive/net6.0/_._": {}
>         }
>       },
>       "ToolsCoreDynLib/1.0.0": {
>         "type": "project",
>         "dependencies": {
>           "Utils": "1.0.0"
>         },
>         "compile": {
>           "bin/placeholder/ToolsCoreDynLib.dll": {}
>         },
>         "runtime": {
>           "bin/placeholder/ToolsCoreDynLib.dll": {}
>         }
>       },
>       "Utils/1.0.0": {
>         "type": "project",
>         "framework": ".NETCoreApp,Version=v8.0",
>         "dependencies": {
>           "System.Collections.Immutable": "8.0.0"
>         },
>         "compile": {
>           "bin/placeholder/Utils.dll": {}
>         },
>         "runtime": {
>           "bin/placeholder/Utils.dll": {}
>         }
>       }
>     }
>   },
>   "libraries": {
>     "System.Collections.Immutable/8.0.0": {
>       "sha512": "AurL6Y5BA1WotzlEvVaIDpqzpIPvYnnldxru8oXJU2yFxFUy3+pNXjXd1ymO+RA0rq0+590Q8gaz2l3Sr7fmqg==",
>       "type": "package",
>       "path": "system.collections.immutable/8.0.0",
>       "files": [
>         ".nupkg.metadata",
>         ".signature.p7s",
>         "Icon.png",
>         "LICENSE.TXT",
>         "PACKAGE.md",
>         "THIRD-PARTY-NOTICES.TXT",
>         "buildTransitive/net461/System.Collections.Immutable.targets",
>         "buildTransitive/net462/_._",
>         "buildTransitive/net6.0/_._",
>         "buildTransitive/netcoreapp2.0/System.Collections.Immutable.targets",
>         "lib/net462/System.Collections.Immutable.dll",
>         "lib/net462/System.Collections.Immutable.xml",
>         "lib/net6.0/System.Collections.Immutable.dll",
>         "lib/net6.0/System.Collections.Immutable.xml",
>         "lib/net7.0/System.Collections.Immutable.dll",
>         "lib/net7.0/System.Collections.Immutable.xml",
>         "lib/net8.0/System.Collections.Immutable.dll",
>         "lib/net8.0/System.Collections.Immutable.xml",
>         "lib/netstandard2.0/System.Collections.Immutable.dll",
>         "lib/netstandard2.0/System.Collections.Immutable.xml",
>         "system.collections.immutable.8.0.0.nupkg.sha512",
>         "system.collections.immutable.nuspec",
>         "useSharedDesignerContext.txt"
>       ]
>     },
>     "ToolsCoreDynLib/1.0.0": {
>       "type": "project",
>       "path": "../ToolsCoreDynLib/ToolsCoreDynLib.vcxproj",
>       "msbuildProject": "../ToolsCoreDynLib/ToolsCoreDynLib.vcxproj"
>     },
>     "Utils/1.0.0": {
>       "type": "project",
>       "path": "../Utils/Utils.csproj",
>       "msbuildProject": "../Utils/Utils.csproj"
>     }
>   },

**Comment by jeffkl** [comment]
@AlexBaratheon restore in Visual Studio works subtly different.  Have you tried setting `EnableManagedPackageReferenceSupport` in your VCXPROJ?

https://devblogs.microsoft.com/cppblog/announcing-nuget-packagereference-support-for-c-cli-msbuild-projects-targeting-net-core/

I wonder if the C++ project system is causing this if that property is not set.

**Comment by zivkan** [comment]
Also in the obj directory is a `{project}.g.nuget.dgspec.json` file. Can you provide a diff of that file?  It's contains NuGet's "inputs", so that will tell us what information NuGe is getting differently from the C++ project system in VS compared to direct from MSBuild on the CLI.

**Comment by AlexBaratheon** [comment]
@zivkan 
Diff of the `ApplicationUtils.csproj.nuget.dgspec.json`:
![2024-10-18 08_39_25-TOR-56138  KST-7106 Der locked_packages json fehle - OneNote](https://github.com/user-attachments/assets/5cfde50b-3a4a-4a9f-bff9-044195cc130c)
![2024-10-18 08_41_13-ApplicationUtils csproj nuget dgspec json VS build and ApplicationUtils csproj n](https://github.com/user-attachments/assets/5e01c4d9-7ac8-45c4-8b49-b4665380767f)


@jeffkl I closed VS, added `EnableManagedPackageReferenceSupport` to the C++/CLI project, started VS and tried to build the solution. Thats what I got:
![2024-10-18 08_26_15-ProjectAssetJsonTests - Microsoft Visual Studio LTSC 17 8](https://github.com/user-attachments/assets/0cd89f91-584e-404e-89bb-e1629106c37c)

I am not sure where the version `10.0.17134.0` comes from, because I could not find it in the `.vcxproj`-file or VS settings:
![2024-10-18 08_49_19-ToolsCoreDynLib Property Pages](https://github.com/user-attachments/assets/6e039924-53c3-4255-974c-2af445da5886)

Second thing I tried was running nuget restore SOLUTION with the inserted `EnableManagedPackageReferenceSupport`. Now I get a third version of the project.assets.json file which is very similar to the one from the VS build. Only with 6 additonal lines.

Diff of the `project.assets.json`: 
VS build vs nuget restore with `EnableManagedPackageReferenceSupport` in the C++/CLI project
![2024-10-18 09_39_22-TOR-56138  KST-7106 Der locked_packages json fehle - OneNote](https://github.com/user-attachments/assets/bb05f0e3-bc32-44b2-b901-a836d6e6dce0)
![2024-10-18 09_42_33-TOR-56138  KST-7106 Der locked_packages json fehle - OneNote](https://github.com/user-attachments/assets/c3b532d8-dc3a-4758-907c-151527f36451)

**Comment by jeffkl** [comment]
@AlexBaratheon That's great, those differences now are probably because you have a newer version of nuget.exe and an older version of Visual Studio.  You're running Visual Studio 17.8 and nuget.exe 6.9, but nuget.exe 6.8 shipped with Visual Studio 17.8.

I would suggest not restoring with `nuget.exe` at all and instead running `msbuild.exe /t:Restore <ProjectPath>` from a Visual Studio Command Prompt window for that instance of Visual Studio.  That will ensure that command-line restores are using the same version as Visual Studio.  But if you need, nuget.exe 6.8 is at https://dist.nuget.org/win-x86-commandline/v6.8.1/nuget.exe

And of course, I recommend running the latest Visual Studio (currently 17.11) if possible.

**Comment by AlexBaratheon** [comment]
@jeffkl Good point, we should not use different nuget versions. I downgraded my nuget.exe from 6.9 to 6.8, to have the version matching to VS, but still with the same results.

On our servers we do not want to start Visual Studio to run `msbuild.exe /t:Restore <ProjectPath>` from the VS Command Promt. The nuget restore and thereby generation of `project.assets.json` needs to be executed from powershell scripts, to be useful for us.

- I also tried out `msbuild.exe /t:Restore SOLUTION` from a command prompt with VS closed. It generates the same project.assets.json as `nuget restore SOLUTION`. 
- Just to see if the issue is resolved on the newest versions, I tried it on another computer. Here with the newest NuGet Version `6.11.1.2` and Visual Studio  `17.11.5` the issue is reproducible.

We sadly can not run the latest Visual Studio in our company, because it is not an LTS version.
